mars/dataframe/utils.py

```diff
@@ -1424,3 +1431,65 @@ def _concat_chunks(merge_chunks: List[ChunkType], output_index: int):
     else:
         params["nsplits"] = (tuple(n_split), df_or_series.nsplits[1])
     return new_op.new_tileable(df_or_series.op.inputs, kws=[params])
+
+
+def clean_up_func(op):
+    closure_clean_up_bytes_threshold = int(
+        os.getenv("MARS_CLOSURE_CLEAN_UP_BYTES_THRESHOLD", 10**4)
+    )
+    if closure_clean_up_bytes_threshold == -1:  # pragma: no cover
+        return
+    ctx = get_context()
+    if ctx is None:
+        return
+    # Before PR #3165 is merged, func cleanup is temporarily disabled under ray task mode.
+    # https://github.com/mars-project/mars/pull/3165
+    if isinstance(ctx, RayExecutionContext):
+        logger.warning("Func cleanup is currently disabled under ray task mode.")
+        return
+    # Note: Vineyard internally uses `pickle` which fails to pickle
+    # cell objects and corresponding functions.
+    if vineyard is not None:
+        storage_backend = ctx.get_storage_info()
+        if storage_backend.get("name", None) == "vineyard":
+            logger.warning(
+                "Func cleanup is currently disabled when vineyard is used as storage backend."
+            )
+            return
+
+    func = op.func
+    if hasattr(func, "__closure__") and func.__closure__ is not None:
+        counted_bytes = 0
+        for cell in func.__closure__:
+            counted_bytes += sys.getsizeof(cell.cell_contents)
+            if counted_bytes >= closure_clean_up_bytes_threshold:
+                op.need_clean_up_func = True
+                break
+    # Note: op.func_key is set to not None only when op.need_clean_up_func is True.
+    if op.need_clean_up_func:
+        # Note: when used in ray task mode, data key is ray.ObjectRef which is returned
+        # after func being put into storage.
+        # if isinstance(ctx, RayExecutionContext):
+        #     op.func_key = ctx.storage_put(None, op.func)
+        # Note: when used in normal mode, data key is logic key while the returned data_info
+        # is not used as key.
+        if isinstance(ctx, ThreadedServiceContext):
+            assert (
+                op.logic_key is not None
+            ), "Logic key wasn't calculated before cleaning up func."
+            op.func_key = op.logic_key
+            ctx.storage_put(op.func_key, op.func)
```

Comment:
Can you return a storage object key from the `storage_put`? Then, we don't need to check instances here.